### PR TITLE
Map tokens to ui config

### PR DESCRIPTION
**Pull request by GB27247**

### CODE REVIEW

File: frontend/pages/settings/personal-access-tokens/new.tsx [hunk 1]
```diff
@@ -18,79 +18,68 @@ import {
...
 
-export default function NewToken() {
   const theme = useTheme()
   const [description, setDescription] = useState('')
   const [isAllModels, setIsAllModels] = useState(true)
   const [selectedModels, setSelectedModels] = useState<EntrySearchResult[]>([])
-  const [isAllActions, setIsAllActions] = useState(true)
-  const [selectedActions, setSelectedActions] = useState<TokenActionKeys[]>(actionOptions)
+  const [isAllActions, setIsAllActions] = useState(false)
+  const [selectedActions, setSelectedActions] = useState<string[]>([])
```
Comment (ARADDCC012) [inline frontend/pages/settings/personal-access-tokens/new.tsx:46]: Do we no longer want to default to all actions being selected?

File: frontend/pages/settings/personal-access-tokens/new.tsx [hunk 2]
```diff
@@ -110,9 +99,26 @@ export default function NewToken() {
           />
         )
       }),
-    [selectedActions],
+    [selectedActions, TokenWriteAction, TokenReadAction],
   )
 
+  if (isUiConfigLoading) {
```
Comment (ARADDCC012) [inline frontend/pages/settings/personal-access-tokens/new.tsx:105]: I'd recommend moving the loading and error handlers to just above your main return block at the bottom. Error handling should also come before loading.
```
  if (isUiConfigError) {
    return <MessageAlert message={isUiConfigError.info.message}></MessageAlert>
  }
  
  if (isUiConfigLoading) {
    return <Loading />
  }
  
  return (
     ...
   )
```

File: frontend/pages/settings/personal-access-tokens/new.tsx [hunk 3]
```diff
@@ -18,79 +18,68 @@ import {
...
   const modelsAutocompletePlaceholder = useMemo(() => {
     if (isAllModels) return 'All models selected'
     if (selectedModels.length) return ''
     return 'Select models'
   }, [isAllModels, selectedModels.length])
 
-  const isGenerateButtonDisabled = useMemo(
-    () => !description || !(isAllModels || selectedModels.length) || !selectedActions.length,
-    [description, isAllModels, selectedModels.length, selectedActions.length],
+  const [TokenReadAction, TokenWriteAction] = actionOptions.reduce<Record<string, string>[]>(
```
Comment (ARADDCC012) [inline frontend/pages/settings/personal-access-tokens/new.tsx:57]: This should probably be in a `useMemo`

File: frontend/pages/settings/personal-access-tokens/new.tsx [hunk 4]
```diff
@@ -18,79 +18,68 @@ import {
...
 
-  const { models, isModelsLoading, isModelsError } = useListModels('model')
-
   const modelsAutocompletePlaceholder = useMemo(() => {
     if (isAllModels) return 'All models selected'
     if (selectedModels.length) return ''
     return 'Select models'
   }, [isAllModels, selectedModels.length])
 
-  const isGenerateButtonDisabled = useMemo(
```
Comment (ARADDCC012) [inline frontend/pages/settings/personal-access-tokens/new.tsx:82]: I think this logic is still needed. We don't want a user to be able to create a token that doesn't work with any models/actions.

**Review by ARADDCC012 — DISMISSED** [review]

File: frontend/pages/settings/personal-access-tokens/new.tsx (hunk 1 above)
Comment (GB27247) [reply frontend/pages/settings/personal-access-tokens/new.tsx:46]: This is pretty standard in other apps. One reason is there's a further barrier to make less restrictive tokens, but I'm happy to hear suggestions.

File: frontend/pages/settings/personal-access-tokens/new.tsx [hunk 5]
```diff
@@ -18,79 +18,72 @@ import {
...
+  const { uiConfig, isUiConfigLoading, isUiConfigError } = useGetUiConfig()
 
-const getActionName = (action: TokenActionKeys) => {
-  const [name, _kind] = action.split(':')
-  return name
-}
+  const tokenActions = uiConfig?.tokenActions || {}
 
-const actionOptions = Object.values(TokenAction)
+  const actionOptions = Object.keys(tokenActions)
```
Comment (ARADDCC012) [inline frontend/pages/settings/personal-access-tokens/new.tsx:39]: Minor, but as `uiConfig.tokenActions` is a decent size object I think it's worth `useMemo`ing this. You'll also have to useMemo `tokenActions` too to make that a stable dependency.

File: frontend/pages/settings/personal-access-tokens/new.tsx (hunk 1 above)
Comment (ARADDCC012) [reply frontend/pages/settings/personal-access-tokens/new.tsx:46]: I think with the improvements being made to the token action selection UI in a future PR, I think this makes sense to do. Happy for you to leave as is.

File: backend/src/utils/config.ts [hunk 6]
```diff
@@ -187,4 +190,5 @@ export interface Config {
 }
 
 const config: Config = _config.util.toObject()
+config.ui.tokenActions = TokenDescriptions
```
Comment (JR40159) [inline backend/src/utils/config.ts:193]: I think this is a clear move away from how the UI config is used and having things like this where the backend appends to the configuration that is provided by the configuration file could easily add confusion in the future. I think having to do this shows that this token information isn't suitable for the UI config. The UI config contains settings that can differ per deployment and need to be shared between the frontend and the backend. The token available actions and their descriptions fulfil the latter but not the former. Instead I'd suggest a simple token endpoint that returns an object of Bailo's available token actions with their descriptions so clients like the UI can display that information to users

File: backend/src/models/Token.ts [hunk 7]
```diff
@@ -29,15 +29,32 @@ export const TokenActions = {
   ImageRead: 'image:read',
   ImageWrite: 'image:write',
 
-  SchemaRead: 'schema:read',
   SchemaWrite: 'schema:write',
+} as const
+
+export const TokenDescriptions = {
```
Comment (JR40159) [inline backend/src/models/Token.ts:35]: I would prefer these two objects to be combined into one. 
```
{
  SchemaWrite: { 
    id: 'schema:write',
    description: 'Grants permissions to upload and modify schemas for administrators.'
  },
  ....
}
```

**Review by JR40159 — CHANGES_REQUESTED** [review]

**Comment by ARADDCC012** [comment]
https://github.com/gchq/Bailo/pull/1332#discussion_r1662388249 is outstanding

**Review by ARADDCC002 — APPROVED** [review]

**Pull request merged**